enhancements/installer/cloud-custom-dns.md

```diff
@@ -0,0 +1,658 @@
+for `api`, `api-int` and `*.apps` resolution, OpenShift will have to provide a
+self-hosted in-cluster DNS solution. This would allow the customer to configure
+their DNS solution post-install. The current proposal focuses on this approach.
+
+OpenShift will start static CoreDNS pods to provide DNS resolution for API,
+Internal API and Ingress services that are essential for cluster creation.
+In order for a master or worker node to come online, they need the
+machine config server via the api-int domain. So, it is essential for the
+api-int domain to be resolvable on the bootstrap node. The static CoreDNS
+pod started on the bootstrap node needs entries for just api-int resolution
+and the static CoreDNS pod started on the control plane nodes needs to be
+able to resolve api, api-int and *.apps domains. Although not identical, this
+approach leverages learnings from OpenShift's approach to providing in-cluster 
+DNS for on-prem platforms.
+
+After cluster deployment is completed, the customer will update their external
+DNS solution with the same assigned LB IP addresses used for the configuration
+of the internal CoreDNS instance. OpenShift will not delete the CoreDNS pod
+even after cluster installation completes.
+
+If the user successfully configures their external DNS service with api,
+api-int and *.apps services, then they could optionally delete the in-cluster
+CoreDNS pod and the cluster is expected to function fully as expected. This is
+a completely optional step with this design. If the customer does configure
+their custom DNS solution and leave the in-cluster CoreDNS pod running, all
+in-cluster entities will end up using the CoreDNS pod's DNS services and all
+out of cluster requests will be handled by the external DNS.
+
+### Workflow Description
+
+1. To use the custom DNS solution, an OpenShift administrator enables it
+via a config option in the install-config. Supplying this config would
+cause the Installer and the Ingress Operator to skip configuring the cloud
+DNS with entries for API, Internal API and Ingress Services.
+
+2. The Infrastructure CR is updated to indicate that the custom DNS and not
+the cloud default DNS is in use.
+
+3. The Installer and the Ingress Operator continue to configure the cloud LBs
+on the public subnets of Public Hosted Zones for `api` and `*.apps`. They will
+continue to configure cloud LBs on the private subnets of Private Hosted Zones
+for `api-int` and `*.apps` for in-cluster access to the Ingress service.
+
+4. After the Installer uses the cloud specific terraform providers to create
+the LBs for API and API-Int, it will add the LB DNS Names of these LB to a [ConfigMap](https://kubernetes.io/docs/concepts/configuration/configmap/). This ConfigMap is
+only created when the custom DNS feature is enabled. This ConfigMap gets
+appended to the Ignition file created by the Installer. Let us call this
+the `lbConfigforDNS` ConfigMap.
```

Comment:
We had explored that option in https://github.com/openshift/enhancements/pull/1276 but ran into some implementation issues within the Installer.

The Installer generates all its manifests, adds them to the bootstrap ignition which is written to an s3 bucket (in the case of AWS) before terraform is started to create and configure cloud infrastructure. So, the Infrastructure manifest has already been written to the bootstrap ignition before the LBs are created via terraform. We found it easier to create a new configmap, append it to the bootstrap ignition and re-write to the s3 bucket than updating the Infrastructure CR that has already been written to the bootstrap ignition file.

Secondly, we don't expect the customer to interact with the configmap at all and for the LB information within it to change (no operator is monitoring the LB values today). It is meant to be a simple mechanism to pass the LB DNS information to MCO from the Installer.

Fwiw, we haven't completely given up on finding a way to update a manifest already written to the bootstrap ignition. Also, another operator (say MCO) could potentially read this configmap and update the Infrastructure CR.

Reply:
From an API perspective, I would much rather see the installer update the status of the infra object (I appreciate the challenges you've outlined) than use a configmap. Configmaps have no validation and aren't real APIs. Having something in cluster change that value based on the configmap value creates a confused deputy style problem.

It's not a blocker per se, but it would help me sleep at night if we could update the infrastructure status directly, or some other in cluster API rather than a configmap

Reply:
@patrickdillon is opposed to updating the Infrastructure CR within the Installer due to the amount of surgery needed on the bootstrap ignition file. I am not sure if the Installer changes needed to remove terraform make things better. Even if they do, that won't be available until a later release.
If the Infrastructure can be updated with the data in ConfigMap by a component other than the Installer, I am open to that. I already explored MCO as an option early on and that doesn't work either. Any other options seem viable?

Reply:
> @patrickdillon is opposed to updating the Infrastructure CR within the Installer due to the amount of surgery needed on the bootstrap ignition file.

I would rather we use the infra CR rather than a CM. I do have reservations with tweaking the contents of bootstrap ignition during infrastructure provisioning, but if we have a good way of doing it I'm open to it. From what I have seen that approach is going against the grain, so we would be better off pursuing something that fits more naturally.

> If the Infrastructure can be updated with the data in ConfigMap by a component other than the Installer, 

1. Installer - we've talked about having the installer itself update the in-cluster object once the API becomes available. This has some small downsides, but I think is a decent bridge to consider. I know this is  blocked until we figure out how the installer can communicate to the API without public DNS.
2. Controller (running on bootstrap) - It seems like the best option is to have a controller monitoring the AWS API for the LB details and then updating the infra CR. I don't think we've considered this much, but from my perspective the biggest issue here would be getting the credentials for the controller. CCO was recently removed from bootkube (perhaps we just put it back, I dunno).

Reply:
Thinking about some alternative solutions...

Looking at the code that applies the manifests, it doesn't use Apply, but rather Create and UpdateStatus. It appears to me that we could add a _second_ Infra CR manifest containing only the Status, and the [Spec would be ignored because the CR already exists](https://github.com/openshift/library-go/blob/master/pkg/assets/create/creater.go#L181-L187), but the [Status would be updated](https://github.com/openshift/library-go/blob/master/pkg/assets/create/creater.go#L213-L229). That would appear to offer the best bits of both the ConfigMap and Infra CR options.

In the baremetal-operator we have some code to help with CAPI pivoting. On the first reconcile only (i.e. when the Status subresource isn't created yet) if it sees a particular annotation it copies the contents into the status (and then later deletes the annotation). Perhaps if we used some sort of similar one-time arrangement then we would avoid the confused deputy problem.

If this is only used by MCO on the bootstrap, does it need to be passed through the k8s API at all? The next list item says that we will "provide the ConfigMap as an optional parameter" (not sure what this means - we provide the name/namespace of it?). Couldn't we provide e.g. a filename to read it from as an optional parameter instead?

If all this is happening only on the bootstrap, and enabled by the optional arg we're passing to MCO only in the bootstrap, does this question even matter? MCO in the running cluster will pay no heed to this ConfigMap. During bootstrapping we completely control the input.

Reply:
> If this is only used by MCO on the bootstrap, does it need to be passed through the k8s API at all? The next list item says that we will "provide the ConfigMap as an optional parameter" (not sure what this means - we provide the name/namespace of it?). Couldn't we provide e.g. a filename to read it from as an optional parameter instead?
@zaneb Yes, we do provide a filename.  See https://github.com/openshift/installer/pull/7662/files. I'll update to clarify.

Reply:
Why wrap it in a ConfigMap and add it to the manifests directory at all then? It could be any old JSON file.

Reply:
> Why wrap it in a ConfigMap and add it to the manifests directory at all then? It could be any old JSON file.
@zaneb we wanted to treat it like an asset generated by the Installer which it is.

Reply:
Going back to @JoelSpeed 's original question. If the LB config were to be added to the Infra CR, what would be the best place to add it? PlatformStatus ? As of now, we don't expect the value to change once the LBs have been configured by the Installer (can change with the upcoming CAPI based changes to the Installer.)

Reply:
Another option would be for MCO to read the configMap and update Infrastructure CR with these values. This was initially thought of as not a possibility because MCO did not own the Infrastructure resource. But, recent discussions seem promising.